src/dxso/dxso_compiler.cpp

```diff
@@ -2070,18 +2070,22 @@ namespace dxvk {
         DxsoRegMask srcMask(true, true, true, false);
         auto vec3 = emitRegisterLoad(src[0], srcMask);
 
-        DxsoRegisterValue dot = emitDot(vec3, vec3);
-        dot.id = m_module.opInverseSqrt (scalarTypeId, dot.id);
+        // No need for emitDot, either both arguments or none are zero.
+        // mul_zero has the same result as ieee mul.
+        uint32_t dot = m_module.opDot(scalarTypeId, vec3.id, vec3.id);
+        dot = m_module.opInverseSqrt(scalarTypeId, dot);
         if (m_moduleInfo.options.d3d9FloatEmulation == D3D9FloatEmulation::Enabled) {
-          dot.id = m_module.opNMin        (scalarTypeId, dot.id,
-            m_module.constf32(FLT_MAX));
+          dot = m_module.opNMin(scalarTypeId, dot, m_module.constf32(FLT_MAX));
         }
 
-        // r * rsq(r . r);
-        result.id = m_module.opVectorTimesScalar(
-          typeId,
-          emitRegisterLoad(src[0], mask).id,
-          dot.id);
+        std::array<uint32_t, 3> dotReplicated = {dot, dot, dot};
+
+        DxsoRegisterValue vecLength;
```

Comment:
It's not really the length. Call it something like `invVecLength` instead.

Reply:
names should be better now.